Update Docker intro (#38947)
dotnet/docs: docs/core/docker/introduction.md

# Introduction to .NET and Docker

> Containers are one of the most popular ways for deploying and hosting cloud applications, with tools like [Docker](https://www.docker.com/), [Kubernetes](https://kubernetes.io/), and [Podman](https://podman.io/). Many developers choose containers because it is straightforward to package an app with its required dependencies and then to get that app to reliably run on any container host. There is entensive support for [using .NET with containers](https://devblogs.microsoft.com/dotnet/category/containers/).

Reviewer: ```suggestion
Containers are one of the most popular ways for deploying and hosting cloud applications, with tools like [Docker](https://www.docker.com/), [Kubernetes](https://kubernetes.io/), and [Podman](https://podman.io/). Many developers choose containers because it is straightforward to package an app with its required dependencies and then to get that app to reliably run on any container host. There is extensive support for [using .NET with containers](https://devblogs.microsoft.com/dotnet/category/containers/).
```

> Containers are one of the most popular ways for deploying and hosting cloud applications, with tools like [Docker](https://www.docker.com/), [Kubernetes](https://kubernetes.io/), and [Podman](https://podman.io/). Many developers choose containers because it is straightforward to package an app with its required dependencies and then to get that app to reliably run on any container host. There is entensive support for [using .NET with containers](https://devblogs.microsoft.com/dotnet/category/containers/).

Reviewer 2: ```suggestion
Containers are one of the most popular ways for deploying and hosting cloud applications, with tools like [Docker](https://www.docker.com/), [Kubernetes](https://kubernetes.io/), and [Podman](https://podman.io/). Many developers choose containers because it's straightforward to package an app with its dependencies and get that app to reliably run on any container host. There's extensive support for [using .NET with containers](https://devblogs.microsoft.com/dotnet/category/containers/).
```

Docker provides a great [overview](https://docs.docker.com/engine/docker-overview/) of containers. [Docker Desktop: Community Edition](https://www.docker.com/products/docker-desktop) is a good tool to use for using containers on developer desktop machine.

## .NET images

> Official .NET Docker images are published to the [Microsoft Artifact Registry](https://mcr.microsoft.com/) and are discoverable on the [Docker Hub](https://hub.docker.com/_/microsoft-dotnet/). There are [runtime images](https://mcr.microsoft.com/product/dotnet/aspnet/) for production and [SDK images](https://mcr.microsoft.com/product/dotnet/sdk/) for building your code, for Linux (Alpine, Debian, Ubuntu) and Windows. For more information, see [.NET container images](container-images.md).

Reviewer: ```suggestion
Official .NET container images are published to the [Microsoft Artifact Registry](https://mcr.microsoft.com/) and are discoverable on the [Docker Hub](https://hub.docker.com/_/microsoft-dotnet/). There are [runtime images](https://mcr.microsoft.com/product/dotnet/aspnet/) for production and [SDK images](https://mcr.microsoft.com/product/dotnet/sdk/) for building your code, for Linux (Alpine, Debian, Ubuntu, Mariner) and Windows. For more information, see [.NET container images](container-images.md).
```

> .NET images are regularly updated, whevever a new .NET patch is published or when an operating system base image is updated.

Reviewer 3: ```suggestion
.NET images are regularly updated whenever a new .NET patch is published or when an operating system base image is updated.
```

> [Chiseled container images](https://devblogs.microsoft.com/dotnet/announcing-dotnet-chiseled-containers/) are Ubuntu container images with only the few components required by the .NET runtime. These images are ~100MB smaller than the regular Ubuntu images and have fewer [CVEs](https://www.cve.org/) since they have fewer components. In particular, they do not container a shell or package manager, which significantly improves their security profile. They also include a [non-root user](https://devblogs.microsoft.com/dotnet/securing-containers-with-rootless/) and are configured with that user enabled.

Reviewer: ```suggestion
[Chiseled container images](https://devblogs.microsoft.com/dotnet/announcing-dotnet-chiseled-containers/) are Ubuntu container images with only the few components required by the .NET runtime. These images are ~100MB smaller than the regular Ubuntu images and have fewer [CVEs](https://www.cve.org/) since they have fewer components. In particular, they do not contain a shell or package manager, which significantly improves their security profile. They also include a [non-root user](https://devblogs.microsoft.com/dotnet/securing-containers-with-rootless/) and are configured with that user enabled.
```

> [Chiseled container images](https://devblogs.microsoft.com/dotnet/announcing-dotnet-chiseled-containers/) are Ubuntu container images with only the few components required by the .NET runtime. These images are ~100MB smaller than the regular Ubuntu images and have fewer [CVEs](https://www.cve.org/) since they have fewer components. In particular, they do not container a shell or package manager, which significantly improves their security profile. They also include a [non-root user](https://devblogs.microsoft.com/dotnet/securing-containers-with-rootless/) and are configured with that user enabled.

Reviewer 2: ```suggestion
[Chiseled container images](https://devblogs.microsoft.com/dotnet/announcing-dotnet-chiseled-containers/) are Ubuntu container images with a minimal set of components required by the .NET runtime. These images are ~100 MB smaller than the regular Ubuntu images and have fewer [CVEs](https://www.cve.org/) since they have fewer components. In particular, they don't contain a shell or package manager, which significantly improves their security profile. They also include a [non-root user](https://devblogs.microsoft.com/dotnet/securing-containers-with-rootless/) and are configured with that user enabled.
```

## Building container images

> You can build a container image with a **Dockerfile** or rely on the [.NET SDK to produce an image](https://devblogs.microsoft.com/dotnet/announcing-builtin-container-support-for-the-dotnet-sdk/). For samples on building images, see [dotnet/dotnet-docker](https://github.com/dotnet/dotnet-docker/blob/main/samples/README.md) and [dotnet/sdk-container-builds](https://github.com/dotnet/sdk-container-builds).

Reviewer 2: ```suggestion
You can build a container image with a **Dockerfile** or rely on the [.NET SDK to produce an image](publish-as-container.md). For samples on building images, see [dotnet/dotnet-docker](https://github.com/dotnet/dotnet-docker/blob/main/samples/README.md) and [dotnet/sdk-container-builds](https://github.com/dotnet/sdk-container-builds).
```

The following example demonstrates building and running a container image in a few quick steps.

```bash
$ dotnet new webapp -o webapp
$ cd webapp/

> $ dotnet publish -p:PublishProfile=DefaultContainer

Reviewer: This will only work with .NET 8. I know that you can set the context of the docs to be for a specific .NET version. I don't know how that is configured wrt these markdown files.

Reviewer 3: This'll work for web projects out of the box from 7.0.300 onwards, but that's a nuance that probably isn't worth noting here :) In addition, we should possibly move away from `-p:PublishProfile=DefaultContainer` to `/t:PublishContainer` instead - folks seem to understand MSBuild Targets a bit more than the Publish Profile concept, and the target-form works better with solution-level publishing.

Reviewer: Right, but what about the people using .NET 6? This needs to be excluded from .NET 6 context or clarified in the text.

$ docker ps
CONTAINER ID   IMAGE     COMMAND               CREATED              STATUS              PORTS                                       NAMES
7c7ad33409e5   webapp    "dotnet webapp.dll"   About a minute ago   Up About a minute   0.0.0.0:8000->8080/tcp, :::8000->8080/tcp   jovial_shtern
$ docker kill 7c7ad33409e5
```

[`docker init`](https://www.docker.com/blog/docker-desktop-4-23/) is a new option for developers wanting to use Dockerfiles.

## Ports

[Port mapping](https://docs.docker.com/network/#published-ports) is a key part of using containers. Ports must be published outside the container in order to respond to external web requests. [ASP.NET Core container images changed in .NET 8](https://learn.microsoft.com/dotnet/core/compatibility/containers/8.0/aspnet-port) to listen on port `8080`, by default. .NET 6 and 7 listen on port `80`.

In the prior example with `docker run`, the host port `8000` is mapped to the container port `8080`. Kubernetes works in a similar way.

The `ASPNETCORE_HTTP_PORTS`, `ASPNETCORE_HTTPS_PORTS`, and `ASPNETCORE_URLS` environment variables can be used to configure this behavior.

## Users

> All .NET 8+ images include the `app` user. Chiseled images are configured to use this user. The SDK container builds feature (demonstrated above) also configures images to use this user. In all other scenarios, this user can be set manually, for example with the `USER` **Dockerfile** instruction. If an image has been configured with `app` and commands need to run as `root`, the then `USER` instruction can be used to set to the user to `root`.

Reviewer: ```suggestion
All .NET 8+ images include the `app` user. Chiseled images are configured with this user enabled by default. The SDK container builds feature (demonstrated above) also configures images with this user enabled by default. In all other scenarios, this user can be set manually, for example with the `USER` **Dockerfile** instruction. If an image has been configured with `app` and commands need to run as `root`, then the `USER` instruction can be used to set to the user to `root`.
```

> All .NET 8+ images include the `app` user. Chiseled images are configured to use this user. The SDK container builds feature (demonstrated above) also configures images to use this user. In all other scenarios, this user can be set manually, for example with the `USER` **Dockerfile** instruction. If an image has been configured with `app` and commands need to run as `root`, the then `USER` instruction can be used to set to the user to `root`.

Reviewer: > SDK container builds feature

This reads awkwardly if you interpret "builds" as a verb.
